AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Endpoint/EndpointClient.swift

```diff
@@ -225,6 +244,47 @@ actor EndpointClient: EndpointClientBehaviour {
     nonisolated private func getOptOut(from endpointProfile: PinpointEndpointProfile) -> String {
         return endpointProfile.isOptOut ? EndpointClient.Constants.OptOut.all : EndpointClient.Constants.OptOut.none
     }
+    
+    private static func migrateLegacyKeyValueStore(_ userDefaults: UserDefaultsBehaviour, _ keychain: KeychainStoreBehavior, _ archiver: AmplifyArchiverBehaviour) {
+        if let endpointProfileData = userDefaults.data(forKey: Constants.endpointProfileKey) {
+            do {
+                try keychain.set(endpointProfileData, key: Constants.endpointProfileKey)
+                userDefaults.removeObject(forKey: Constants.endpointProfileKey)
+            } catch {
+                log.error("Unable to migrate Analytics key-value store for key \(Constants.endpointProfileKey)")
```

Comment:
I've just realized something: We are attempting to migrate everything to the Keychain on this function and we are removing any existing data from UserDefaults while doing so. 

However, if the `keychain.set` operations fail, we keep the data on UserDefaults (which is good), but then we only attempt to retrieve their values from the keychain (see lines 66, 71. 151 and 164).

This basically means that if we have stored values in UserDefaults and the migration to the Keychain fails, we will behave as if we didn't have any stored value and we might even end up with two different values in each storage after a while

Reply:
That's a very good point.  I've added static getter methods that will fallback to Userdefault lookup if keychain is empty.   I've left the implementation in `EndpointClient` since it's only a couple static functions.  I've also made similar updates to the AWS Pinpoint framework in aws-ios-sdk, please also review the [PR 4223](https://github.com/aws-amplify/aws-sdk-ios/pull/4223) with the latest updates.